src/vm/functions/principals.rs

```diff
@@ -38,13 +53,97 @@ pub fn special_is_standard(
         _ => return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, owner).into()),
     };
 
-    let address_is_mainnet = version == C32_ADDRESS_VERSION_MAINNET_MULTISIG
-        || version == C32_ADDRESS_VERSION_MAINNET_SINGLESIG;
-    let address_is_testnet = version == C32_ADDRESS_VERSION_TESTNET_MULTISIG
-        || version == C32_ADDRESS_VERSION_TESTNET_SINGLESIG;
+    let address_is_mainnet = version_matches_mainnet(version);
+    let address_is_testnet = version_matches_testnet(version);
     let context_is_mainnet = env.global_context.mainnet;
 
     Ok(Value::Bool(
         (address_is_mainnet && context_is_mainnet) || (address_is_testnet && !context_is_mainnet),
     ))
 }
+
+pub fn native_principal_parse(principal: Value) -> Result<Value> {
+    let (version_byte, hash_bytes) = match principal {
+        Value::Principal(PrincipalData::Standard(StandardPrincipalData(version, bytes))) => {
+            (version, bytes)
+        }
+        Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier { issuer, name })) => {
+            (issuer.0, issuer.1)
+        }
+        _ => {
+            return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, principal).into())
+        }
+    };
+
+    if !version_matches_mainnet(version_byte) && !version_matches_testnet(version_byte) {
+        return Err(CheckErrors::InvalidVersionByte.into());
```

Comment:
I think this should be a `RuntimeErrorType`. In general, `CheckErrors` are errors that _should_ be caught by a type checker, and if they aren't, they _invalidate_ the transaction (meaning that the transactions can't even be included in a block).

Reply:
Is `(principal-parse)` supposed to terminate the transaction if given an invalid version byte?  That sounds needlessly egregious to me.

Also, if all we're doing here is *parsing* principals, then why make an unexpected version byte into an error at all?  Isn't there a separate method for testing the version byte to see if it's standard?  There's no guarantee that the set of version bytes we support today will be the *only* version bytes supported in the future.

Reply:
Could we make it so that at least `(principal-parse)` returns `(response (tuple (....)) uint)` like the other methods that could possibly fail already do?

Reply:
Again.. how to decide? I'm fine to do either on both questions.